In [1]:
# Import the modules
import numpy as np
import pandas as pd
#from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report
from imblearn.over_sampling import RandomOverSampler


---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df = pd.read_csv("lending_data.csv")  # The instructions on Canvas specify that the notebook and the csv file
                                              # are to be put in the same folder. There is no need to import Path.

# Review the DataFrame
lending_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [3]:
# Review the data
lending_df.info()       # all data is numeric with no missing values


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2   borrower_income   77536 non-null  int64  
 3   debt_to_income    77536 non-null  float64
 4   num_of_accounts   77536 non-null  int64  
 5   derogatory_marks  77536 non-null  int64  
 6   total_debt        77536 non-null  int64  
 7   loan_status       77536 non-null  int64  
dtypes: float64(3), int64(5)
memory usage: 4.7 MB


### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_df['loan_status']

# Separate the X variable, the features
X = lending_df.drop('loan_status', axis=1)

In [5]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [6]:
# Review the X variable DataFrame
X.head(10)

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   
5    10100.0          7.438            50600        0.407115                4   
6    10300.0          7.490            51100        0.412916                4   
7     8800.0          6.857            45100        0.334812                3   
8     9300.0          7.096            47400        0.367089                3   
9     9700.0          7.248            48800        0.385246                4   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  
5                 1       20600  
6                 1       21100  
7                 0       15100  
8                 0       17400  
9                 0       18800

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [7]:
# Check the balance of our target values
y.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [8]:
# Import the train_test_learn module
    # Done at the top of the file

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
    # Done at the top of the file

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
testing_predictions = logistic_regression_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Print the balanced_accuracy score of the model

print(f"Balanced_accuracy score: {balanced_accuracy_score(y_test, testing_predictions)}")

Balanced_accuracy score: 0.9442676901753825


In [12]:
# Generate a confusion matrix for the model
test_matrix = confusion_matrix(y_test, testing_predictions)

In [13]:
# Print the classification report for the model
test_matrix

array([[18679,    80],
       [   67,   558]], dtype=int64)

In [14]:
# Create and save the testing classification report
testing_report = print(classification_report(y_test, testing_predictions, target_names=["0: healthy loan","1: high-risk loan"]))

# Print the testing classification report
testing_report

                   precision    recall  f1-score   support

  0: healthy loan       1.00      1.00      1.00     18759
1: high-risk loan       0.87      0.89      0.88       625

         accuracy                           0.99     19384
        macro avg       0.94      0.94      0.94     19384
     weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** <br>
Despite an accuracy score of 0.99 for the model, the high imbalance between the number of healthy loans (75036) vs. defaulted loans (2500) causes a bias in the model, and the f1-score for high-risk loans is only 0.88 despite using the stratification when splitting the data (2500 x0.25 = 625 support entries). <br>
This situation creates a number of false negatives that might be not acceptable for the financial institution.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [15]:
# Import the RandomOverSampler module form imbalanced-learn
    # Done at the top of the file

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_train_res, y_train_res = ros.fit_resample(X_train, y_train)


# Had to reinstall imbalance-learn in conda in my dev environment
# conda install -c conda-forge imbalanced-learn
# from https://imbalanced-learn.org/stable/install.html

In [16]:
# Count the distinct values of the resampled labels data
y_train_res.value_counts()

loan_status
0    56277
1    56277
Name: count, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [18]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model_res = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using the resampled training data
logistic_regression_model_res.fit(X_train_res, y_train_res)

# Make a prediction using the testing data
testing_predictions_res = logistic_regression_model_res.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [19]:
# Print the balanced_accuracy score of the model 
print(f"Balanced_accuracy score: {balanced_accuracy_score(y_test, testing_predictions_res)}")

Balanced_accuracy score: 0.9959744975744975


In [22]:
# Generate a confusion matrix for the model
test_matrix_res = confusion_matrix(y_test, testing_predictions_res)
test_matrix_res

array([[18668,    91],
       [    2,   623]], dtype=int64)

In [21]:
# Create and save the testing classification report
testing_report_res = print(classification_report(y_test, testing_predictions_res, target_names=["0: healthy loan","1: high-risk loan"]))

# Print the testing classification report
testing_report_res

                   precision    recall  f1-score   support

  0: healthy loan       1.00      1.00      1.00     18759
1: high-risk loan       0.87      1.00      0.93       625

         accuracy                           1.00     19384
        macro avg       0.94      1.00      0.96     19384
     weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** <br>By oversampling the high-risk loan data for rebalancing prior to fitting the model, the number of False Negatives has dropped dramatically from 67 to 2.<br>
The number of False Positives has increased slightly from 80 to 91, which has a lower impact on the financial institution (missed opportunity but not a direct loss).<br>
Overall, the f1-score for the high-risk loans has increased from 0.88 to 0.93 after oversampling, which leads to a model accuracy of 1.0 